src/lib/dfp/fill-external-slots.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * External slots are a special type of dynamic slot.
+ *
+ * They are not fixed (aka SSR) or dynamic (aka injected by spacefinder).
+ * They are placed on the page by a non-standard route for example in a thrasher or some
+ * other async process that adds the slot at an unknown time but still expects the
+ * commercial runtime to fulfill the slot.
+ *
+ * The extra logic in addition to dynamic slots covers when:
+ * - the commercial runtime loads before the slot so we wait for a custom event
+ * - the commercial runtime loads after the slot so we fill the slot immediately
+ *
+ * External slots will not work from within a restricted iframe such as a
+ * cross-origin or safeframe iframe.
+ */
+
+import { fillDynamicAdSlot } from './fill-dynamic-advert-slot';
+
+const EXTERNAL_SLOT_PREFIX = 'dfp-ad--external';
+
+type ExternalSlotCustomEventDetail = {
+	slotId: string;
+};
+
+type ExternalSlotCustomEvent = {
+	detail: ExternalSlotCustomEventDetail;
+};
+
+const isCustomEvent = (event: Event): event is CustomEvent => {
+	return 'detail' in event;
+};
+
+const fillExternalSlots = () => {
+	const externalSlots = document.querySelectorAll<HTMLElement>(
+		`[id^=${EXTERNAL_SLOT_PREFIX}]`,
+	);
+	return [...externalSlots].map((slot) => fillDynamicAdSlot(slot, false));
+};
+
+const createSlotFillListener = () => {
+	document.addEventListener('gu.commercial.slot.fill', (event: Event) => {
```

Comment:
Do we want this to be listening specifically for external slot fill events? (i.e. should we give it a more specific name?)

```suggestion
	document.addEventListener('gu.commercial.external.slot.fill', (event: Event) => {
```

Reply:
I thought it might be useful to make this generic so that if we extend this in future any kind of slot could be picked up and fulfilled. Does that seem okay?